test/e2e/network/networking_perf.go

```diff
@@ -18,145 +18,279 @@ package network
+
+	ginkgo.It(fmt.Sprintf("should run iperf2"), func() {
+		readySchedulableNodes, err := e2enode.GetReadySchedulableNodes(f.ClientSet)
+		framework.ExpectNoError(err)
+
+		familyStr := ""
+		if framework.TestContext.ClusterIsIPv6() {
+			familyStr = "-V "
+		}
+
+		serverPodsListOptions := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", labelKey, serverLabelValue),
 		}
 
-		framework.Logf("Reading all perf results to stdout.")
-		framework.Logf("date,cli,cliPort,server,serverPort,id,interval,transferBits,bandwidthBits")
+		// Step 1: set up iperf2 server -- a single pod on any node
+		serverDeployment, err := iperf2ServerDeployment(f.ClientSet, f.Namespace.Name, framework.TestContext.ClusterIsIPv6())
+		framework.ExpectNoError(err, "deploy iperf2 server deployment")
 
-		// Extra 1/10 second per client.
-		iperfTimeout := smallClusterTimeout + (time.Duration(expectedCli/10) * time.Second)
-		iperfResults := &IPerfResults{}
+		_, err = iperf2ServerService(f.ClientSet, f.Namespace.Name)
+		framework.ExpectNoError(err, "deploy iperf2 server service")
 
-		iperfClusterVerification := f.NewClusterVerification(
-			f.Namespace,
-			framework.PodStateVerification{
-				Selectors:   iperfClientPodLabels,
-				ValidPhases: []v1.PodPhase{v1.PodSucceeded},
-			},
-		)
-
-		pods, err2 := iperfClusterVerification.WaitFor(expectedCli, iperfTimeout)
-		if err2 != nil {
-			framework.Failf("Error in wait...")
-		} else if len(pods) < expectedCli {
-			framework.Failf("IPerf restuls : Only got %v out of %v, after waiting %v", len(pods), expectedCli, iperfTimeout)
-		} else {
-			// For each builds up a collection of IPerfRecords
-			iperfClusterVerification.ForEach(
-				func(p v1.Pod) {
-					resultS, err := framework.LookForStringInLog(f.Namespace.Name, p.Name, "iperf-client", "0-", 1*time.Second)
-					if err == nil {
-						framework.Logf(resultS)
-						iperfResults.Add(NewIPerf(resultS))
-					} else {
-						framework.Failf("Unexpected error, %v when running forEach on the pods.", err)
-					}
-				})
+		// Step 2: set up iperf2 client daemonset
+		//   initially, the clients don't do anything -- they simply pause until they're called
+		_, err = iperf2ClientDaemonSet(f.ClientSet, f.Namespace.Name)
+		framework.ExpectNoError(err, "deploy iperf2 client daemonset")
+
+		clientPodsListOptions := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", labelKey, clientLabelValue),
 		}
-		fmt.Println("[begin] Node,Bandwidth CSV")
-		fmt.Println(iperfResults.ToTSV())
-		fmt.Println("[end] Node,Bandwidth CSV")
 
-		for ipClient, bandwidth := range iperfResults.BandwidthMap {
-			framework.Logf("%v had bandwidth %v.  Ratio to expected (%v) was %f", ipClient, bandwidth, expectedBandwidth, float64(bandwidth)/float64(expectedBandwidth))
+		// Make sure the server is ready to go
+		framework.ExpectNoError(e2edeployment.WaitForDeploymentComplete(f.ClientSet, serverDeployment), "unable to wait for server deployment")
+
```

Comment:
we have to wait until the endpoints are ready:

```go
err = wait.Poll(2*time.Second, largeClusterTimeout, func() (done bool, err error) {
	listOptions := metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", discoveryv1beta1.LabelServiceName, serverServiceName)}
	esList, err := f.ClientSet.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).List(context.TODO(), listOptions)
	framework.ExpectNoError(err, "Error fetching EndpointSlice for Service %s/%s", f.Namespace.Name, serverServiceName)

	if len(esList.Items) == 0 {
		framework.Logf("EndpointSlice for Service %s/%s not found", f.Namespace.Name, serverServiceName)
		return false,nil
	}
        return true, nil
})
framework.ExpectNoError(err, "unable to wait for endpoints for the iperf service")

Reply:
👍 thanks for the pointer @aojea, I'll take a look!

Reply:
okay just pushed this up, let's see how it does ... :excited: 😄